Enable running bpfd as systemd service

Billy99 opened the pull request:
As part of this PR, eased the path to allow bpfd to run as a systemd service.

The following changes were made:
* Added a new `setup.sh` script to manage all the installation.
* Added a new `user.sh` script to handle creating a bpfd and bpfctl user
  and creating the `/etc/bpfd/` and a new a `/etc/bpfctl/` directory. Also
  create the associated sub-folders and assign user and user group to
  the files and directories. Also copy over a default `bpfd.toml` and
  `bpfctl.toml` file.
* Added a new `install.sh` script to copy the binaries to `/usr/sbin/.`
  and assign user and user group to the files. Also copy a default
  `bpfd.service` file to `/usr/lib/systemd/system/`.
* Modified the `certificates.sh` script to be called by `setup.sh`.
* The `bpfd.service` limits the set of capabilities bpfd service has.

Changes in behavior:
* Default `bpfd.toml` and `bpfctl.toml` files are copied over.
* Moved `/etc/bpfd.toml` to `/etc/bpfd/bpfd.toml`.
* Mover `/etc/bpfd/bpfctl.toml` to `/etc/bpfct/bpfctl.toml`.
* `certificates.sh` is no longer called directly. All certs creation
  goes through `setup.sh`.

See `tutorial.md` and `./scripts/setup.sh help` for full description
of how everything now works.

Signed-off-by: Billy McFall <22157057+Billy99@users.noreply.github.com>

Billy99 commented:
Updated the directory structure as follows:

Original:
```
$ sudo tree /etc/bpfd/
/etc/bpfd.toml
/etc/bpfd/
├── bpfctl.toml
├── certs
│   ├── bpfctl
│   │   ├── bpfctl.key
│   │   └── bpfctl.pem
│   ├── bpfd
│   │   ├── bpfd.key
│   │   └── bpfd.pem
│   ├── ca
│   │   ├── ca.key
│   │   └── ca.pem
│   └── gocounter
│       ├── gocounter.key
│       └── gocounter.pem
└── programs.d
    └── pass.toml
```
New:
```
$ sudo tree /etc/bpfd/
/etc/bpfd/
├── bpfd.toml
├── certs
│   ├── bpfd
│   │   ├── bpfd.key
│   │   └── bpfd.pem
│   └── ca
│       ├── ca.key
│       └── ca.pem
└── programs.d
    └── pass.toml

$ sudo tree /etc/bpfctl/
/etc/bpfctl/
├── bpfctl.toml
└── certs
    ├── bpfctl
    │   ├── bpfctl.key
    │   └── bpfctl.pem
    └── gocounter
        ├── gocounter.key
        └── gocounter.pem

$ sudo tree /usr/lib/systemd/system/
/usr/lib/systemd/system/
 :
├──bpfd.service
 :

$ sudo tree /usr/sbin/
/usr/sbin/
 :
├── bpfctl
├── bpfd
 :
```

Billy99 commented:
As far which capabilities bpfd requires, I ran `sudo /usr/share/bcc/tools/capable`, then ran through a bunch of actions with `bpfd` and `bpfctl`. Here is the output and a summary: 
https://gist.github.com/Billy99/888b295cdd4ec63105bc8b3899497c90

The set of capabilities needed are:
CAP_BPF
CAP_DAC_READ_SEARCH
CAP_NET_ADMIN
CAP_PERFMON
CAP_SYS_ADMIN
CAP_SYS_RESOURCE

The goal was to drop capabilities on threads that didn't need them. So far, all attempts to drop capabilities on thread have cause bpfd to stop working. 
```
$ ps -ef | grep bpfd
root      214723  202880  0 Aug18 pts/7    00:00:00 journalctl -f -u bpfd
bpfd      229703       1  0 12:59 ?        00:00:00 /usr/sbin/bpfd
bmcfall   229773  200636  0 13:00 pts/2    00:00:00 grep --color=auto bpfd

$ ps -T -p 229703
    PID    SPID TTY          TIME CMD
 229703  229703 pts/3    00:00:00 bpfd
 229703  229704 pts/3    00:00:00 tokio-runtime-w
 229703  229705 pts/3    00:00:00 tokio-runtime-w
 229703  229706 pts/3    00:00:00 tokio-runtime-w
 229703  229707 pts/3    00:00:00 tokio-runtime-w
```

astoycos commented on bpfctl/src/main.rs:
```diff
@@ -12,6 +12,8 @@ mod config;
 use config::config_from_file;
 use tonic::transport::{Certificate, Channel, ClientTlsConfig, Identity};
 
+const DEFAULT_BPFCTL_CONFIG_PATH: &str = "/etc/bpfctl/bpfctl.toml";
```
I wonder should we be putting global constants like this in the `bpfd-common` package?

Billy99 replied:
I would like a common place. If I get around to selinux policy, I need to know all the files we manage. I moved to the top of the file as a place holder.

astoycos commented on bpfd/src/server/bpf.rs:
```diff
@@ -267,7 +267,12 @@ impl<'a> BpfManager<'a> {
                     let cmsg = ControlMessage::ScmRights(&fds);
-                    sendmsg(sock, &iov, &[cmsg], MsgFlags::empty(), Some(&sock_addr)).unwrap();
+                    if let Err(result) =
+                        sendmsg(sock, &iov, &[cmsg], MsgFlags::empty(), Some(&sock_addr))
```
Something I found out the other day is that you could use `sendmsg(sock, &iov, &[cmsg], MsgFlags::empty(), Some(&sock_addr)).map_error(BpfdError::NotAuthorized)?` 

If you wanted to one line it, but you'd lose that `info!` message

dave-tucker replied:
[`map_err`](https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err) is totally fine,
the info log message shouldn't be needed...

we probably need to log these errors when they occur from elsewhere in bpfd (i.e in `rpc.rs` before we return them to the client)

Billy99 replied:
I had a scenario where `gocounter` created a socket and had access to it and `bpfd` did not (I was still working through all the permissions). Without the info log, it just fails silently. I don't know how many other scenarios that will occur, but I think it is helpful. I do need to change the returned error. I just picked one, but I'll create a new BpfdError::SendError.

astoycos commented on bpfd/src/server/mod.rs:
```diff
@@ -54,12 +56,16 @@ pub async fn serve(
         .serve(addr);
 
     tokio::spawn(async move {
+        drop_linux_capabilities();
```
Was doing this twice on purpose?

astoycos commented on docs/admin/configuration.md:
```diff
@@ -3,8 +3,10 @@ Configuration
+bpfd expects a configuration file to be present at `/etc/bpfd/bpfd.toml`.
 If no file is found, defaults are assumed.
+The `scripts/setup.sh` commands will copy a default version of the file to the correct location
+from `scripts/bpfd.toml`, which can then be overwritten if needed.
```
So now we essentially will never use the rust defaults right?

Billy99 replied:
If the script is run, we essentially will never use the rust defaults. But we can't be guaranteed the script is run.

astoycos commented on docs/admin/configuration.md:
```diff
@@ -48,12 +50,14 @@ proceed_on = ["pass", "dispatcher_return"]
   ca_cert = "/etc/bpfd/certs/ca/ca.pem"
-  cert = "/etc/bpfd/certs/bpfctl/bpfctl.pem"
-  key = "/etc/bpfd/certs/bpfctl/bpfctl.key"
+  cert = "/etc/bpfctl/certs/bpfctl/bpfctl.pem"
```
nit: can we get rid of the nested bpfctl dir now and just have `/etc/bpfctl/certs/bpfctl.pem`

Billy99 replied:
I currently have `/etc/bpfctl/certs/gocounter/` as well.

astoycos commented on docs/admin/tutorial.md:
```diff
@@ -1,53 +1,81 @@
+  See [Basic Mode](#basic-mode).
+* `bpfctl` without `sudo` - Run `bpfd` in foreground or background and straight from
+  build directory, but also create the `bpfd` or `bpfctl` users and user groups.
+  So if `bptctl` user group is added to a given user, `sudo` is no longer needed when executing `bptctl` commands.
```
```suggestion
  So if `bpfctl` user group is added to a given user, `sudo` is no longer needed when executing `bptctl` commands.
```

astoycos commented on docs/admin/tutorial.md:
```diff
@@ -1,53 +1,81 @@
-## Step 2: Start bpfd
+### Step 2: Setup `bpfd` environment
+
+`bpfd` uses mTLS for mutual authentication.
```
```suggestion
`bpfd` uses mTLS for mutual authentication with clients.
```

astoycos commented on docs/admin/tutorial.md:
```diff
@@ -1,53 +1,81 @@
+
+`bpfd` uses mTLS for mutual authentication.
+In order to run and interact with `bpfd`, certificates must be created.
+Scripts to create self-sign certificates can be used in a development environment.
```
```suggestion
Scripts to create a local certificate authority which is used to sign certs for bpfd and bpfctl can be used in a development environment.
```

I think is the right terminology, since "technically" they aren't self signed certs

Billy99 replied:
ack

astoycos commented on docs/admin/tutorial.md:
```diff
@@ -171,13 +200,148 @@ xdp_mode: skb
+
+### Step 3: Start `bpfd`
+
+To manage `bpfd` as a systemd service, use `systemclt`:
```
```suggestion
To manage `bpfd` as a systemd service, use `systemctl`:
```

Billy99 replied:
ack

Billy99 commented on docs/admin/tutorial.md:
```diff
@@ -1,53 +1,81 @@
 # Tutorial
 
```
I have one. Maybe it would help to bold the 3 modes:
```suggestion
# Tutorial

This tutorial will show you how to use `bpfd`.
There are several ways to launch and interact with `bpfd` and `bpfctl`:

* **Basic Mode** - Run `bpfd` in foreground or background and straight from build directory.
  No `bpfd` or `bpfctl` users are created so always need `sudo` when executing `bptctl` commands.
  See [Basic Mode](#basic-mode).
* **`bpfctl` without `sudo`** - Run `bpfd` in foreground or background and straight from
  build directory, but also create the `bpfd` or `bpfctl` users and user groups.
  So if `bptctl` user group is added to a given user, `sudo` is no longer needed when executing `bptctl` commands.
  See [`bpfctl` without `sudo`](#bpfctl-without-sudo).
* **Systemd Service** - Run "bpfd" as a systemd service.
  See [Systemd Service](#systemd-service).

```

Billy99 replied:
I'll also rename:
"Basic Mode" as "Priviledged Mode"
"`bpfctl` without `sudo`" as "Unpriviledged Mode"

dave-tucker commented on bpfd/src/main.rs:
```diff
@@ -11,7 +11,11 @@ use nix::{
+const DEFAULT_BPFD_STATIC_PROGRAM_DIR: &str = "/etc/bpfd/programs.d";
+const BPFD_ENV_LOG_LEVEL: &str = "RUST_LOG";
+
+#[tokio::main(flavor = "multi_thread", worker_threads = 1)]
```
No :smile: 

Main thread = BPF thread.
Spawned thread = RPC thread.

The `current_thread` executor would use a single OS thread to run BOTH those tasks. Not what we want.

The `multi_thread` executor spawns a thread per CPU and schedules work across them. This is cool, because the `RPC` thread will quite likely call `spawn` also, BUT, those workers inherit the capabilities of the main thread unless we drop them.

The correct option is probably to manually configure the runtime using https://docs.rs/tokio/1.20.1/tokio/runtime/struct.Builder.html and use `on_thread_start` to drop permissions

Billy99 replied:
Fixed

Billy99 commented:
I addressed all the comments except around error handling.
* Updated the scripts such that there are still two users (bpfd and bpfctl), but one user group (bpfd)
* Fixed the change regarding the number of threads.
* Add additional commit to allow gocounter example to run with limited privileges (no sudo)
* Fixed comments in documents

Billy99 commented:
Should the user `bpfctl` and the related directory `/etc/bpfctl/` be renamed something for all bpfd clients? `bpfd-client`?

astoycos commented on bpfd/src/server/rpc.rs:
```diff
@@ -194,8 +194,7 @@ impl Loader for BpfdLoader {
-        match res {
+        match resp_rx.await {
             Ok(_) => Ok(Response::new(reply)),
             Err(e) => Err(Status::aborted(format!("{}", e))),
```
Maybe we make a new issue for this @Billy99 Unless you'd like to tackle it here

Billy99 replied:
I'm here, might as well do it now.

Billy99 commented:
Should the user `bpfctl` and the related directory `/etc/bpfctl/` be renamed something for all bpfd clients? `bpfd-client`?

> Same, just wasn't sure of the name.

astoycos commented:
I think `/etc/bpf-clients` makes sense @Billy99  But I'm fine with `/etc/bpfd-client` as well

dave-tucker commented:
Going to merge this one as-is. I have some comments, but we can address those in a follow up (will open issues).
Then we can get #102 in

dave-tucker commented:
Great work @Billy99 :medal_sports:

astoycos commented:
+1 thanks @Billy99!
